pkg/inject/inject.go

```diff
@@ -0,0 +1,585 @@
+}
+
+// ParseMeta extracts metadata from bytes.
+// It returns false if the workload's payload is empty
+func (conf *ResourceConfig) ParseMeta(bytes []byte) (bool, error) {
+	if err := yaml.Unmarshal(bytes, &conf.meta); err != nil {
+		return false, err
+	}
+	if err := yaml.Unmarshal(bytes, &conf.objMeta); err != nil {
+		return false, err
+	}
+	return conf.objMeta.ObjectMeta != nil, nil
+}
+
+// GetPatch returns the JSON patch containing the proxy and init containers specs, if any
+func (conf *ResourceConfig) GetPatch(bytes []byte) ([]byte, []Report, error) {
+	report := newReport(conf)
+	log.Infof("working on %s %s..", strings.ToLower(conf.meta.Kind), conf.objMeta.Name)
+
+	if err := conf.parse(bytes); err != nil {
+		return nil, []Report{report}, err
+	}
+
+	var patch *Patch
+	if conf.meta.Kind == "Pod" {
+		patch = NewPatchPod()
+	} else {
+		patch = NewPatchDeployment()
+	}
+
+	// If we don't inject anything into the pod template then output the
+	// original serialization of the original object. Otherwise, output the
+	// serialization of the modified object.
+	if conf.podSpec != nil {
+		metaAccessor, err := k8sMeta.Accessor(conf.obj)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// The namespace isn't necessarily in the input so it has to be substituted
+		// at runtime. The proxy recognizes the "$NAME" syntax for this variable
+		// but not necessarily other variables.
+		identity := k8s.TLSIdentity{
+			Name:                metaAccessor.GetName(),
+			Kind:                strings.ToLower(conf.meta.Kind),
+			Namespace:           "$" + PodNamespaceEnvVarName,
+			ControllerNamespace: conf.globalConfig.GetLinkerdNamespace(),
+		}
+
+		report.update(conf)
+		if conf.shouldInject(report) {
+			conf.injectPodSpec(patch, identity)
+			conf.injectObjectMeta(patch)
+		}
+	} else {
+		report.UnsupportedResource = true
+	}
+
+	patchJSON, err := json.Marshal(patch.patchOps)
+	log.Debugf("patch: %s", patchJSON)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return patchJSON, []Report{report}, nil
+}
+
+// KindInjectable returns true if the resource in conf can be injected with a proxy
+func (conf *ResourceConfig) KindInjectable() bool {
+	for _, kind := range injectableKinds {
+		if conf.meta.Kind == kind {
+			return true
+		}
+	}
+	return false
+}
+
+func (conf *ResourceConfig) getFreshWorkloadObj() interface{} {
+	var obj interface{}
+
+	switch conf.meta.Kind {
+	case "Deployment":
+		obj = v1beta1.Deployment{}
+	case "ReplicationController":
+		obj = v1.ReplicationController{}
+	case "ReplicaSet":
+		obj = v1beta1.ReplicaSet{}
+	case "Job":
+		obj = batchv1.Job{}
+	case "DaemonSet":
+		obj = v1beta1.DaemonSet{}
+	case "StatefulSet":
+		obj = appsv1.StatefulSet{}
+	case "Pod":
+		obj = v1.Pod{}
+	}
+	return obj
+}
+
+// JSONToYAML is a replacement for the same function in sigs.k8s.io/yaml
+// that does conserve the field order as portrayed in k8s' api structs
+func (conf *ResourceConfig) JSONToYAML(bytes []byte) ([]byte, error) {
```

Comment:
@siggy The original problem was that when applying the JSON patch with `jsonpatch.Apply()`, that lib uses a map under the hood to unmarshal the original JSON before applying the patch, which disrupted the order.
We're keeping that, and the following step is to convert the patched JSON into YAML. I can't recall if k8s' `yaml.JSONToYAML()` kept the ordering, but that doesn't matter since we already have a disordered JSON at this point.
So our `conf.JSONToYAML()` uses its knowledge of the workload's type to pass it into k8s' `yaml.Unmarshal()` to build that typed workload object fully filled (actually I'm gonna change this and use the std lib `json.Unmarshal()` instead) which kinda "restores" the order according to the struct, and then I'm finally ready to transform it into YAML.